File: ZeroTier One/NetworkInfoCell.m
//
//  NetworkInfoCell.m
//  ZeroTier One
//
//  Created by Grant Limberg on 8/7/16.
//  Copyright © 2016 ZeroTier, Inc. All rights reserved.
//

#import "NetworkInfoCell.h"
#import "ServiceCom.h"
#import "ShowNetworksViewController.h"
#import "Network.h"

@implementation NetworkInfoCell

- (void)drawRect:(NSRect)dirtyRect {
    [super drawRect:dirtyRect];
    
    // Drawing code here.
}

- (IBAction)onConnectCheckStateChanged:(NSButton*)sender
{
    if(sender.state == NSOnState) {
        [self joinNetwork:self.networkIdField.stringValue];
    }
    else {
        [self leaveNetwork:self.networkIdField.stringValue];
    }
}

- (IBAction)deleteNetwork:(NSButton*)sender;
{
    [self leaveNetwork:self.networkIdField.stringValue];
    [self.parent deleteNetworkFromList:self.networkIdField.stringValue];
}

- (IBAction)onAllowStatusChanged:(NSButton*)sender
{
    [self joinNetwork:self.networkIdField.stringValue];
}

- (void)joinNetwork:(NSString*)nwid
{
    NSError *error = nil;
    [[ServiceCom sharedInstance] joinNetwork:nwid
                                allowManaged:(self.allowManaged.state == NSOnState)
                                 allowGlobal:(self.allowGlobal.state  == NSOnState)
                                allowDefault:![Network defaultRouteExists:_parent.networkList] && (self.allowDefault.state == NSOnState)
                                       error:&error];

    if (error) {
        // TODO: Display error message
    }
}

- (void)leaveNetwork:(NSString*)nwid
{
    NSError *error = nil;
    [[ServiceCom sharedInstance] leaveNetwork:nwid error:&error];

    if (error) {
        // TODO: Display error message
    }
}

@end

File: ZeroTier One/NetworkInfoCell.swift
//
//  NetworkInfoCell.swift
//  ZeroTier One
//
//  Created by Grant Limberg on 5/24/16.
//  Copyright © 2016 ZeroTier, Inc. All rights reserved.
//

import Cocoa

class NetworkInfoCell: NSTableCellView {

    weak var parent: ShowNetworksViewController!

    @IBOutlet var networkIdField: NSTextField!
    @IBOutlet var networkNameField: NSTextField!

    @IBOutlet var statusField: NSTextField!
    @IBOutlet var typeField: NSTextField!
    @IBOutlet var macField: NSTextField!
    @IBOutlet var mtuField: NSTextField!
    @IBOutlet var broadcastField: NSTextField!
    @IBOutlet var bridgingField: NSTextField!
    @IBOutlet var deviceField: NSTextField!
    @IBOutlet var addressesField: NSTextField!

    @IBOutlet var allowManaged: NSButton!
    @IBOutlet var allowGlobal: NSButton!
    @IBOutlet var allowDefault: NSButton!

    @IBOutlet var connectedCheckbox: NSButton!
    @IBOutlet var deleteButton: NSButton!

    override func awakeFromNib() {
        super.awakeFromNib()
        
    }
    override func drawRect(dirtyRect: NSRect) {
        super.drawRect(dirtyRect)

        // Drawing code here.
    }

    @IBAction func onConnectCheckStateChanged(sender: NSButton) {
        if(sender.state == NSOnState) {
            joinNetwork(networkIdField.stringValue)
        }
        else {
            leaveNetwork(networkIdField.stringValue)
        }
    }

    @IBAction func deleteNetwork(sender: NSButton) {
        leaveNetwork(networkIdField.stringValue)
        parent.deleteNetworkFromList(networkIdField.stringValue)
    }

    func joinNetwork(nwid: String) {
        ServiceCom.joinNetwork(nwid,
                               allowManaged: allowManaged.state == NSOnState,
                               allowGlobal: allowGlobal.state == NSOnState,
                               allowDefault: !defaultRouteExists(parent.networkList) && (allowDefault.state == NSOnState))
    }

    func leaveNetwork(nwid: String) {
        ServiceCom.leaveNetwork(nwid)
    }

    @IBAction func onAllowStatusChanged(sender: NSButton) {
        joinNetwork(networkIdField.stringValue)
    }
}
